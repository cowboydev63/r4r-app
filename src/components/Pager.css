.r4r-pager.r4r-DEFAULT {
    display: flex;
    justify-content: flex-end;
    margin-top: 0.6em;
    margin-bottom: 0.6em;
}

.r4r-pager__count.r4r-DEFAULT {
    display: flex;
    justify-content: flex-end;
    align-items: center;
    margin-right: auto;
}

.r4r-pager__count p:first-child::after {
    content: '|';
    margin: 0 10px;
}

.r4r-pager__nav.r4r-DEFAULT {
    display: flex;
    justify-content: flex-end;
    align-items: center;
}

.pager__num.r4r-DEFAULT, .pager__arrow.r4r-DEFAULT {
    padding: 3px 6px;
    border-radius: 3px;
    font-weight: 400;
    color: #2b7bba;
    display: flex;
    justify-content: center;
    align-items: center;
    cursor: pointer;
    transition: all 150ms linear;
}

.r4r-pager .pager__num.r4r-DEFAULT:focus, .r4r-pager .pager__arrow.r4r-DEFAULT:focus {
    background: #cbe8fc;
}

.r4r-pager.r4r-DEFAULT .pager__num.r4r-DEFAULT, .r4r-pager.r4r-DEFAULT .pager__arrow.r4r-DEFAULT {
    margin: 0 3px;
}

.pager__num.r4r-DEFAULT:hover, .pager__arrow.r4r-DEFAULT:hover {
    background: #cbe8fc;
}

.pager__num.pager__ellipses:hover, .pager__num.pager__num--active:hover {
    background: none;
    cursor: default;
}


.r4r-pager.r4r-DEFAULT .pager__num--active.r4r-DEFAULT {
    background: #2a71a5;
    color: white;
    cursor: default;
    user-select: none;
}