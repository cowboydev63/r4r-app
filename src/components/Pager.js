import React from 'react';
import PropTypes from 'prop-types';
import { Theme } from '../theme';
import PagerCounter from './PagerCounter';
import { 
    keyHandler,
    formatPagerArray,
} from '../utilities';
import '../polyfills/array_fill';
import './Pager.css';

class Pager extends React.PureComponent {
    static propTypes = {
        total: PropTypes.number.isRequired,
        startFrom: PropTypes.number,
        pageSize: PropTypes.number.isRequired,
        resultsSize: PropTypes.number.isRequired,
        onClick: PropTypes.func,
        withCounter: PropTypes.bool,
    }

    static defaultProps = {
        pageSize: 20,
        onClick: () => {},
        withCounter: false,
        total: 0,
        resultsSize: 0,
        startFrom: 1,
    }

    onClick = (from, isCurrent) => () => {
        if(!isCurrent) {
            this.props.onClick({ from })
        }
    }

    renderPages = (total, current) => {
        const pages = formatPagerArray(total, current);
        
        const renderedPages = pages.map((el, idx) => {
            const isCurrent = current === el;
            return el
                ?
                    <Theme
                        element="div"
                        key={ idx } 
                        className={ `pager__num ${ isCurrent ? 'pager__num--active' : ''}`}
                        onClick={ this.onClick(((el - 1) * this.props.pageSize), isCurrent) }
                        onKeyPress={ keyHandler({
                            fn: this.onClick(((el - 1) * this.props.pageSize), isCurrent),
                        })}
                        tabIndex="0"
                        role="link"
                        aria-label={`Results Page ${ el }`}
                    >
                    { el }
                    </Theme>
                :
                    <Theme
                        element="div"
                        key={ idx }
                        className={ `pager__num pager__ellipses`}
                    >
                    ...
                    </Theme>
        })
        
        return renderedPages;
    }

    render(){
        const {
            total,
            startFrom,
            pageSize,
            resultsSize,
        } = this.props;

        if(!total || !resultsSize) {
            return null;
        }

        const isSinglePageOnly = total <= pageSize;
        const isFirstPage = (startFrom + resultsSize) <= pageSize;
        const isLastPage = startFrom >= total - pageSize;
        const totalPages = Math.ceil(total / pageSize);
        const currentPage = Math.ceil((startFrom + 1) / pageSize);
        return (
            !isSinglePageOnly &&
            <Theme element="nav" className="r4r-pager">
                {
                    /* Allow an optional results counter */
                    this.props.withCounter &&
                        <PagerCounter
                            from={ startFrom + 1 }
                            to={ startFrom + resultsSize }
                            total={ total }
                        />
                }
                <Theme element="div" className='r4r-pager__nav'>
                    { 
                        !isFirstPage && 
                            <Theme
                                element="div"
                                className='pager__arrow' 
                                tabIndex="0"
                                onClick={ this.onClick((startFrom - pageSize), false) }
                                onKeyPress={ keyHandler({
                                    fn: this.onClick((startFrom - pageSize), false),
                                })}
                                aria-label="previous results page"
                                role="link"
                            >
                            { '< Previous' }
                            </Theme> 
                    }
                    {
                        this.renderPages(totalPages, currentPage)
                    }
                    { 
                        !isLastPage && 
                            <Theme
                                element="div" 
                                className='pager__arrow' 
                                tabIndex="0"
                                onClick={ this.onClick((startFrom + pageSize), false) }
                                onKeyPress={ keyHandler({
                                    fn: this.onClick((startFrom + pageSize), false),
                                })}
                                aria-label="next results page"
                                role="link"
                            >
                            { 'Next >' }
                            </Theme> 
                    }
                </Theme>
            </Theme>
        )
    }
}


export default Pager;