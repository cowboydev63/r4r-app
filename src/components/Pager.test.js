import React from 'react';
import { shallow } from 'enzyme';
import Pager from './Pager';

describe('Pager Component', () => {
  it('renders without crashing', () => {
    shallow(<Pager />);
  });

  it('renders correctly', () => {
    const wrapper = shallow(<Pager total={100} resultsSize={20} />);
    expect(wrapper).toMatchSnapshot();
  })
})