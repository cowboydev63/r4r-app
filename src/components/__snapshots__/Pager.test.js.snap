// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pager Component renders correctly 1`] = `
<React.Fragment>
  <Theme
    className="r4r-pager__nav"
    element="div"
  >
    <Theme
      aria-label="previous results page"
      className="pager__arrow"
      element="div"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="link"
      tabIndex="0"
    >
      &lt;
    </Theme>
    <Theme
      aria-label="Results Page 1"
      className="pager__num pager__num--active"
      element="div"
      key="0"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="link"
      tabIndex="0"
    >
      1
    </Theme>
    <Theme
      aria-label="Results Page 2"
      className="pager__num "
      element="div"
      key="1"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="link"
      tabIndex="0"
    >
      2
    </Theme>
    <Theme
      aria-label="Results Page 3"
      className="pager__num "
      element="div"
      key="2"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="link"
      tabIndex="0"
    >
      3
    </Theme>
    <Theme
      aria-label="Results Page 4"
      className="pager__num "
      element="div"
      key="3"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="link"
      tabIndex="0"
    >
      4
    </Theme>
    <Theme
      aria-label="Results Page 5"
      className="pager__num "
      element="div"
      key="4"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="link"
      tabIndex="0"
    >
      5
    </Theme>
    <Theme
      aria-label="next results page"
      className="pager__arrow"
      element="div"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="link"
      tabIndex="0"
    >
      &gt;
    </Theme>
  </Theme>
</React.Fragment>
`;
